### PR TITLE
stdlib: Add binary search function

### PR DESCRIPTION
**Pull request by LemonBoy**

### CODE REVIEW

**Comment by andrewrk** [comment]
I think if you add a test that does a binary search looking for a runtime known value, it will reveal a flaw in the API.

I think it would be a simpler to use with a predicate that takes two arguments.

**Comment by SpexGuy** [comment]
Should this be a compile error? In isolation, makeComparisonPred appears to "bake" the runtime variable 'value' into a struct definition, which is not actually something Zig supports.

**Comment by andrewrk** [comment]
zig actually does support this, but only if the value is compile-time known. `type` is a type that requires a compile-time known value, so a function which returns a `type` is always called at compile-time. So all parameters to functions that return `type` are always comptime parameters.

**Comment by LemonBoy** [comment]
> I think it would be a simpler to use with a predicate that takes two arguments.

That way you lose a lot of flexibility. Consider a slice of structures `S` with many fields, sometimes you just want to perform the binary search according to the value of a single field of type `T` and that's easy to do with this API as you just need a predicate that does `math.compare(S.field, key)`.

On the other hand if the `compareFn` is of type `S -> S -> Order` you'd have to build a bogus record of type `S` and you'd still have to write a comparison predicate anyway.

> I think if you add a test that does a binary search looking for a runtime known value, it will reveal a flaw in the API.

Indeed, without closures it becomes quite awkward. I guess we'll have to go for a bsearch-like interface :(

**Comment by LemonBoy** [comment]
Oh, to top everything off it turns out you cannot use `math.order` as predicate as it's too generic:
```
expected type 'fn(u32, u32) std.math.Order', found 'fn(var,var) var'
```

**Comment by andrewrk** [comment]
Can you humor me and add a test that does a binary search looking for a runtime known value?

**Comment by andrewrk** [comment]
> Oh, to top everything off it turns out you cannot use `math.order` as predicate as it's too generic:

hmmmm... I think actually that coercion should work. zig can instantiate the generic function based on the expected type. that can be a separate issue.

File: lib/std/sort.zig [hunk 1]
```diff
@@ -5,6 +5,66 @@ const mem = std.mem;
 const math = std.math;
 const builtin = @import("builtin");
 
+pub fn binarySearch(comptime T: type, key: T, items: []T, comptime compareFn: fn (lhs: T, rhs: T) math.Order) ?usize {
```
Comment (andrewrk) [inline lib/std/sort.zig]: should be `[]const T` right?

File: lib/std/sort.zig [hunk 2]
```diff
@@ -5,6 +5,66 @@ const mem = std.mem;
...
 const builtin = @import("builtin");
 
+pub fn binarySearch(comptime T: type, key: T, items: []const T, comptime compareFn: fn (lhs: T, rhs: T) math.Order) ?usize {
+    if (items.len < 1)
+        return null;
+
+    var left: usize = 0;
+    var right: usize = items.len - 1;
+
+    while (left <= right) {
```
Comment (andrewrk) [inline lib/std/sort.zig]: Should this be `compareFn(left, right) != .gt`?

Comment (LemonBoy) [reply lib/std/sort.zig]: hm? Those are indices

Comment (andrewrk) [reply lib/std/sort.zig]: oops!

**Pull request merged**